### PR TITLE
VAULT-35369: Loaded snapshots CE

### PR DESCRIPTION
**Pull request by miagilepner**
### Description
This PR adds the CE components to support loaded snapshots.
Ent PR: https://github.com/hashicorp/vault-enterprise/pull/8015
RFC: https://go.hashi.co/prd/vlt-045

### TODO only if you're a HashiCorp employee
- [ ] **Backport Labels:** If this fix needs to be backported, use the appropriate `backport/` label that matches the desired release branch. Note that in the CE repo, the latest release branch will look like `backport/x.x.x`, but older release branches will be `backport/ent/x.x.x+ent`.
    - [ ] **LTS**: If this fixes a critical security vulnerability or [severity 1](https://www.hashicorp.com/customer-success/enterprise-support) bug, it will also need to be backported to the current [LTS versions](https://developer.hashicorp.com/vault/docs/enterprise/lts#why-is-there-a-risk-to-updating-to-a-non-lts-vault-enterprise-version) of Vault. To ensure this, use **all** available enterprise labels.
- [x] **ENT Breakage:** If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE
  files.
- [x] **Jira:** If this change has an associated Jira, it's referenced either
  in the PR description, commit message, or branch name.
- [x] **RFC:** If this change has an associated RFC, please link it in the description.
- [x] **ENT PR:** If this change has an associated ENT PR, please link it in the
  description. Also, make sure the changelog is in this PR, _not_ in your ENT PR.

### CODE REVIEW

File: vault/raft_test.go [hunk 1]
```diff
@@ -35,115 +31,3 @@ func TestFormatDiscoveredAddr(t *testing.T) {
 		}
 	}
 }
-
-// TestRaftDirPath verifies that the Raft data directory path is correctly extracted from the storage configuration.
-// It exercises all execution paths in the RaftDataDirPath to ensure that:
-//   - The path is correctly extracted on a happy path.
-//   - In case of an empty path in config, it returns an empty string and true.
-//   - If there are any issues getting the path, it returns an empty string and false.
-func TestRaftDirPath(t *testing.T) {
```
Comment (guygrigsby) [inline vault/raft_test.go:44]: I am curious about why this test is deleted?

**Review by guygrigsby — DISMISSED** [review]

File: vault/raft_test.go (hunk 1 above)
Comment (miagilepner) [reply vault/raft_test.go:44]: Yeah, fair question. I deleted it because the new method (of grabbing the directory from the raft config itself) didn't allow for testing the more complicated edge cases. The test would basically end up being: create a core with raft storage and verify that the path is non-empty. Then create a core with non-raft storage, and verify the path is empty. I can add that test, I just wasn't sure how much value it really adds. That being said, it is better than nothing.

File: physical/raft/snapshot.go [hunk 2]
```diff
@@ -545,12 +544,8 @@ func loadSnapshot(db *bolt.DB, logger log.Logger, snapshotFile io.ReadCloser, pa
...
 				}
 
-				if pathsToFilter != nil {
-					keyToLookUp := []byte(entry.Key)
-					_, _, ok := pathsToFilter.Root().LongestPrefix(keyToLookUp)
-					if ok {
-						continue
-					}
+				if filterKey != nil && filterKey(entry.Key) {
+					continue
```
Comment (biazmoreira) [inline physical/raft/snapshot.go:548]: Are we allowing **any** path to be filtered? I guess it doesn't really make sense to prevent users to filter some system specific paths, right?

**Review by biazmoreira — APPROVED** [review]
I think it's overall good. I just have a couple of questions.

**Review by kubawi — DISMISSED** [review]
LGTM

File: physical/raft/snapshot.go (hunk 2 above)
Comment (miagilepner) [reply physical/raft/snapshot.go:548]: It's a fair point, but I think in the context of this code, there's no knowledge of what the paths are. The physical/raft stuff just thinks of the paths as arbitrary keys, and it should be up to the caller to define what keys should be kept and what shouldn't be.

**Review by kubawi — APPROVED** [review]
LGTM

**Review by victorr — APPROVED** [review]
All good 👍
